broadinstitute/single_cell_portal_core
# Cell filtering by numeric annotations: client library (SCP-5385)

This adds a client library for filtering cells by quantitative dimensions.

### Overview
Cell annotations can be categorical or numeric.  Cell filtering currently supports categorical (i.e. group-based) annotations, where users check and uncheck boxes for each group in an annotation.

Now, we can programmatically filter by numeric annotations as well.  This enables users to easily answer important questions, for example "what is this gene's expression in samples collected very soon after this experimental milestone", or "very soon _or_ very late after it".

The technical design of this cell filtering extension enables rich quantitative queries.  You can apply a variety of operators in any filter, and all apply multiple filters in the same numeric facet.

### Technical examples
Here are some ways you can filter with numbers; see "Tests" for more technical context.

#### Simple numeric filters
* `["equals", 1.3]`
* `["not equals", 1.3]`
* `["greater than", 6]`
* `["greater than or equal to", 6]`
* `["less than", 6]`
* `["less than or equal to", 6]`
* `["between", [5, 42]]` -- inclusive, i.e. 5 <= d <= 42
* `["not between", [5, 42]]` -- inclusive, i.e. !(5 <= d <= 42)
 
#### Compound numeric filters
* `[["between", [95, 100]], ["between", [1200, 1300]]` -- (95 <= d <= 100) or (1200 <= d <= 1300)

Compound numeric filters could be used to e.g.:
* Isolate the peaks of multimodal distributions (as in above concrete example)
* Isolate the tails of distributions

### Video
Here's how it looks!  (My next PR will include a UI for numeric filters controls.)

https://github.com/broadinstitute/single_cell_portal_core/assets/1334561/188da161-0d5f-4f1b-9ac1-0790a25e6e9f

### Test
A new automated test verifies this library extension, and other tests have been updated accordingly.  

Here's how you can manually test it:
1.  Go to the "Cellular and transcriptional diversity in human lactation" study
2.  Ensure Vite is running
3.  In line 430 of `ExploreDisplayTabs.jsx`, uncomment this line:
```
// window.SCP.updateFilteredCells = updateFilteredCells
```
4.  In line 228 of `cell-faceting.js`, uncomment this line:
```
// console.log('# filtered results', results.length)
```
5.  Open Console panel in DevTools and run the following
```
window.SCP.updateFilteredCells({'time_post_partum_days--numeric--study': [['between', [0, 9]]]}) // 0 <= days <= 9
// window.SCP.updateFilteredCells({'time_post_partum_days--numeric--study': [['less than', 10], ['between', [40, 50]], ['greater than or equal to', 100]]}) // days < 10 OR 40 <= days <= 50 OR days >= 100 
// window.SCP.updateFilteredCells({'time_post_partum_days--numeric--study': [['not between', [100, 200]]]}) // !(100 <= days <= 200)
// window.SCP.updateFilteredCells({'time_post_partum_days--numeric--study': [['not equals', 0]]}) // days !== 0
```
6.  Confirm the Console notes: "filtered results 15302"
7.  Restore the branch via e.g. `git checkout .`

### Further details
More context is available in [2024-01-23 SCP demo](https://drive.google.com/file/d/1IYxq8W0_Z706dzdNAxV0-RQvdZ9HRKfb/view?usp=drive_web).  This relates to SCP-5385.
